### ISSUE TITLE
use-after-free in IOCP Reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8870]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#8870|
|Type|defect|
|Created|2016-10-23 00:05:21Z|
|Branch|https://github.com/twisted/twisted/tree/8870-sendrefcount|

The symptom here is that `twisted.web.test.test_distrib.DistribTests.test_largeReturn` and `twisted.web.test.test_distrib.DistribTests.test_largeWrite` _both_ fail intermittently.  Here's a sample failure:

https://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/207/steps/iocp/logs/problems

It's hard to see what's going on here; the issue is that the "actual" value (the one retrieved from doing some real I/O) has, buried in the middle of it:

`xxxxxxxxxxxxxxxxxx\x00\x00(\x82xxxxxxxxxxxxxxxxxxxxx`

<details><summary>Searchable metadata</summary>

```
trac-id__8870 8870
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__None None
branch__8870_sendrefcount 8870-sendrefcount
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1477181121348099 1477181121348099
changetime__1477476929065730 1477476929065730
version__None None
owner__glyph glyph
cc__PenguinOfDoom
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

More possibly-interesting telemetry from the errors:

```
>>> b = file("blub.txt", 'rb')
>>> segments = b.read().split("'")
>>> b.close()
>>> len(segments)
5
>>> map(len, segments)
[33, 655370, 4, 655361, 1]
>>> actual = segments[1]
>>> expected = segments[3]
>>> actual = eval("'" + actual + "'")
>>> expected = eval("'" + expected + "'")
>>> len(actual)
655361
>>> len(expected)
655361
>>> actual.index('\x00')
262113
>>> '\x00' in expected
False
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

is 262113 an interesting number?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Adding PoD in the hopes that he has a guess...

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_4"></a>|@mithrandi commented|
|-|-|

Possibly a duplicate of [#3525](https://github.com/twisted/twisted/issues/3525).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

mithrandi - yeah, I think that this might in fact be the same bug.

I believe that we're seeing different manifestations of it, but the basic issue is that we initiate two concurrent `write` operations with the same buffer, which in some cases just duplicates the data (if they don't actually overlap) or corrupts said buffer (if they do actually overlap).

I'm trying to understand the case that the patch there fixes, though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

I think I'm getting closer to understanding the problem.

The tests which are failing the most regularly now are `twisted.web.test.test_distrib.DistribTests`, specifically `test_largeWrite` and `test_largeReturn`.

Both of these are testing a condition where a large value is being delivered over PB as a series of smaller writes.  Specifically, by being just a _little bit_ over the size limit, it will be broken into two PB messages each.

Due to the "happy" accident of each PB packet being delivered as one `write` call, the optimization which synchronously calls `write` without buffering in the case of an empty buffer, and the fact that this is a loopback connection, the results may well be received as the result of two separate `WSARecv` calls on two separate reactor iterations on the reader's side.

So this is a probable sequence of events:

The first distrib call is received, this triggers a `write` to the HTTP server transport. The reactor ticks over to process its timed call, which triggers `_resumeWriting` and then `doWrite`.

This clears out `_writeScheduled`, so the next `write` call may kick off another `doWrite`.

Let's say that, since it's an OVERLAPPED operation that probably can never complete "immediately", we get a WSA_IO_PENDING (i.e. ERROR_IO_PENDING).  This means there is an "in flight" write-request which may or may not have completed yet.

Now, the second distrib PB message arrives, and it's time to call `write` again.  We do this same song and dance again, (perhaps) before the reactor gets the dequeued event object from our first `write()`.  Now we have two concurrent write completion operations happening concurrently on the same buffer; the buffer doesn't get cleared until `_handleWrite` gets to deal with the `numBytesWritten` value.

This suggests that the patch on [#3525](https://github.com/twisted/twisted/issues/3525) is roughly correct, although it needs a better explanation of _why_ it works :).

(As a super extra bonus, `wsasend.pxi` calls `PyObject_AsReadBuffer` on its `buff` argument, which doesn't document what, if anything, it does to `buff`'s refcount: https://docs.python.org/2.7/c-api/objbuffer.html - This object is a `bytes` object that is immediately GC'd; maybe that's fine.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Experimentally, my previous interpretation of events is not borne out.  Inserting some debug prints, I still see failures on those tests, but the "write in progress" flag is never tripped in `doWrite`.  The code is still wrong in principle, but something else is causing the practical failure.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Tweaking `SEND_LIMIT` to various... interesting... values has interesting results in my local test environment.

It's set to 128k presently, where it fails.

Tuning it down to 4k or so makes the test reliably pass for 100s of runs.

I'm wondering if this is just an artifact of memory pressure making larger buffers get collected faster or something?  Or an underlying bug in IOCP?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Yyeeeeeaaaaaaaaaaaahhhhhhhh, um

Here's a patch that makes it pass for a hundred or so runs here (it's still passing, have to run at the moment, but you get the idea):

```
#!diff
diff --git a/src/twisted/internet/iocpreactor/tcp.py b/src/twisted/internet/iocpreactor/tcp.py
index cbb9e3f..8772e88 100644
--- a/src/twisted/internet/iocpreactor/tcp.py
+++ b/src/twisted/internet/iocpreactor/tcp.py
@@ -85,8 +85,9 @@ class Connection(abstract.FileHandle, _SocketCloser, _AbortingMixin):
         sent is limited to a size of C{self.SEND_LIMIT}.
         """
         writeView = memoryview(buff)
-        return _iocp.send(self.getFileHandle(),
-            writeView[0:self.SEND_LIMIT].tobytes(), evt)
+        holdThisValue = writeView[0:self.SEND_LIMIT].tobytes()
+        evt.hold = holdThisValue
+        return _iocp.send(self.getFileHandle(), holdThisValue, evt)
 
 
     def _closeWriteConnection(self):

```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

OK, so, squirreling it away like that in Python is working around a lifecycle-management bug in C by encoding some higher-level workarounds.  After messing with the cython source for a little bit to address the problem at its source:

```
PASSED (skips=1, successes=14)

Test Pass 3564
```

Feeling pretty confident about this solution now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

PR available at https://github.com/twisted/twisted/pull/573

The new doc-comments (docstrings are not allowed in cython) that I added should explain the lifecycle-management bug that was present before.  Basically, the kernel would be happily WSASend()-ing away its data, and then the Python GC would come along and deallocate the bytes object that held the buffer backing that invocation to WSASend().  If we were unlucky, some new caller would scribble over it.  So, my parenthetical aside about `PyObject_AsReadBuffer` was in fact the root cause.

Normally this sort of refcount bug would be a crasher! However, since it's just something reading an opaque fixed-size buffer, there's no crash, just a read of some trampled-over garbage.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Status is green on github!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_15"></a>|@mithrandi set owner to @glyph|
|-|-|

I don't feel like I'm qualified to verify that there are no ancillary issues here, but as far as my understanding of Cython goes, the explanation for the bug is plausible, and the fix appears to strictly increase the correctness of the code, so +1 please merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Thanks very much for the review!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set status to `closed`|
|-|-|

In changeset adeb31dfcb2d89aefe2bba411164e6adbafa581b
```
#!CommitTicketReference repository="" revision="adeb31dfcb2d89aefe2bba411164e6adbafa581b"
Merge 8870-sendrefcount: fix IOCP use-after-free

Author: glyph

Reviewer: mithrandi

Fixes: #8870

The IOCP reactor no longer transmits the contents of uninitialized memory when
writing large amounts of data.
```
